packages/nimble-components/src/table/models/array-to-tree.ts

```diff
@@ -56,7 +56,7 @@ export function arrayToTree<TData extends TableRecord>(
         }
 
         // add the current item's data to the item in the lookup table
-        lookup[itemId]!.clientRecord = item;
+        lookup[itemId]!.clientRecord = { ...item };
```

Comment:
Could we add a comment here and in the other spots in data manager about why we are doing a shallow clone?

https://github.com/ni/nimble/blob/f24f5ea3938d8ba15e4f896090f64c0d5021c4f7/packages/nimble-components/src/table/models/data-hierarchy-manager.ts#L32

https://github.com/ni/nimble/blob/f24f5ea3938d8ba15e4f896090f64c0d5021c4f7/packages/nimble-components/src/table/models/data-hierarchy-manager.ts#L40

It's not necessary for the arrayToTree algorithm, it's for a side-effect elsewhere, so might be good to call out.

Also the non-hierarchy code path seems to get away with a shallow clone of the top-level array instead of shallow clones of each individual record. 

https://github.com/ni/nimble/blob/f24f5ea3938d8ba15e4f896090f64c0d5021c4f7/packages/nimble-components/src/table/index.ts#L947-L948

Just to confirm, is it necessary for the hierarchy code path to be different and require shallow clones of each record or is there something missed that could avoid that extra work?

Reply:
The short answer is that we need to make a shallow clone of records in all cases that `setData()` is called, and we don't need to make a shallow clone of records in any other cases. Since `arrayToTree` gets called in multiple code paths, I've updated the approach for cloning to clone in `setData()` itself rather than in the data hierarchy manager.